### DISCUSSION TITLE
Why is there no JP support?

### DISCUSSION THREAD

**Discussion by Raqoon** [daye10/dokkan-battle-farm-bot-app#7] [General]
I get the whole SafetyNet thing, but isn't it on the roadmap?

There's 2 bots to my knowledge that has Japan support can't their bypass be added to this?

**Comment by daye10** [comment]
there is a way to pass safetynet, but for that you would have to spend some money and especially devote a lot of time to it. Unfortunately we can't afford that, I'm still a full time student and my teammate also works full time, sorry

**Reply by Raqoon** [reply]
No problem!

Do you have any idea as to what kind of sum/capital is required? And for what especially?

**Reply by daye10** [reply]
It requires some rooted Android phones that can launch Dokkan, and for the amount of phone it depends, the more users there are the more phones will be needed to ensure enough token to bypass the security, I can't give details since I never tried to do it, I just have an idea on how it works. The necessary minimum would be a rooted android phone  that can launch Dokkan to begin somewhere.

that's more or less how 16 works, and for the other web bot, it must also have tokens provided to it in X or Y way, for the web bot, it gets your identifier and the person behind that can literally do whatever he wants with your account if you don't change it.

**Reply by c3rtificate** [reply]
That's how it used to work. I improved the method and moved the actual generation off of the devices. You hit a hard limit after 5 tokens on the same instance a minute. Meanwhile, it is a time-constrained token that expires after a minute or so, it's an extremely inefficient use of devices, hence I just harvested the device-specific information DroidGuard collects and feed it to the coroutine responsible for handling the tokens. Having thousands of users with this method is just not viable you would need an unrealistic amount of devices, and furthermore generate the tokens on demand. 

This is not easy, I personally have spent 9 months of my time trying to understand the DroidGuard VM and the bytecode that accompanies it. The real kicker is that they release silent updates every 2 months or so where the internal layout changes, some due to ASLR and some due to actual source changes. (So if you want to pursue this, be prepared for their scrutiny) 

Anyhow great job updating this relatively outdated source to work with the latest patch!

**Reply by daye10** [reply]
@c3rtificate 
Thank you very much for all these details, we just had some idea of how it worked and how to bypass this security but not that it was so long and complicated.

Anyway I don't think we would have done it because of lack of time and especially lack of means. But it's always good to know the why and the how!

My teammate and I were wondering if there would be a way to contact you because we have some other questions (not about safetynet) ?

Thanks again for the detailed answer!
